Fix pasting unsaved changes as temporary scratch layers

nyalldawson opened the pull request:
Alternative fix to https://github.com/qgis/QGIS/pull/57985, which fixes the underlying issue.

I think these changes are too risky to merge for 3.42.0, so I propose holding them off till 3.44 (and temporarily including https://github.com/qgis/QGIS/pull/57985 as a fix in the meantime)

nyalldawson committed:
Use QgsUnsetValueAttribute for unset fields when creating new features

Instead of a string value of the default value clause (eg "Autonumber"),
initialize the attribute to a QgsUnsetValueAttribute

Refs #38913

nyalldawson committed:
Handle QgsUnsetAttributeValue in text edit widget wrapper

Fixes #38913

nyalldawson commented:
FYI @uclaros

uclaros commented:
Aha! This makes sense :)

Maybe also extend `void TestQgisAppClipboard::copyPaste()` to handle the unsaved edits state?

nyalldawson committed:
Update test

nyalldawson committed:
Move pasteToNewMemoryVector method to QgsClipboard

Slightly reduces code in qgisapp.cpp, and makes this method
testable without requiring QgisApp

nyalldawson committed:
Add test for paste to new scratch layer with unset field values

nyalldawson commented:
@uclaros 
> Maybe also extend void TestQgisAppClipboard::copyPaste() to handle the unsaved edits state?
Done
